File: native/capacitor-nearby/ios/Sources/NearbyPlugin.swift
import Capacitor
import Foundation
import MultipeerConnectivity
import UIKit

/**
 Transport « à proximité » pour iOS, sur MultipeerConnectivity.

 Pendant Apple du plugin Nearby Connections d'Android. Les deux tiennent le
 même contrat TypeScript et **ne se parlent pas entre eux** : Multipeer est
 propre à Apple, Nearby est propre à Google. Ce transport relie donc des
 appareils de même famille, et le BLE reste le seul chemin hors-ligne entre
 iOS et Android. C'est une limite des plateformes, pas du socle.

 ## Correspondance des vocabulaires

 Le contrat parle d'`endpointId` et d'`endpointName` — les termes de Nearby.
 Multipeer, lui, ne connaît que des `MCPeerID`. On indexe donc les pairs par
 leur `displayName`, seul identifiant stable qui traverse le pont Capacitor.
 Ce nom porte le code court de la partie : c'est aussi le seul champ lisible
 pendant la découverte, donc le seul filtre possible avant connexion.

 ## État de vérification

 Compile sur runner macOS. **Jamais exécuté sur appareil** — il faudrait deux
 iPhone et une application signée. Le contrat qu'il honore est en revanche
 couvert par les tests de `packages/transport-nearby`.
 */
@objc(NearbyPlugin)
public class NearbyPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "NearbyPlugin"
    public let jsName = "Nearby"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startAdvertising", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopAdvertising", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startDiscovery", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopDiscovery", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestConnection", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "acceptConnection", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "send", returnType: CAPPluginReturnPromise),
    ]

    private var peerId: MCPeerID?
    private var session: MCSession?
    private var advertiser: MCNearbyServiceAdvertiser?
    private var browser: MCNearbyServiceBrowser?

    /// Pairs vus, indexés par leur nom affiché — l'`endpointId` du contrat.
    private var found: [String: MCPeerID] = [:]
    /// Invitations reçues, en attente de la décision du socle.
    private var pendingInvitations: [String: (Bool, MCSession?) -> Void] = [:]

    // MARK: - Disponibilité

    @objc func isAvailable(_ call: CAPPluginCall) {
        // Multipeer existe sur tout iPhone depuis iOS 7 ; ce qui manque
        // éventuellement, c'est l'autorisation réseau local — qu'on ne peut pas
        // interroger, seulement constater à l'usage.
        call.resolve([
            "available": true,
            "reason": "",
        ])
    }

    // MARK: - Annonce (hôte)

    @objc func startAdvertising(_ call: CAPPluginCall) {
        guard let serviceId = call.getString("serviceId"),
            let endpointName = call.getString("endpointName")
        else {
            call.reject("serviceId ou endpointName manquant")
            return
        }

        let me = MCPeerID(displayName: endpointName)
        peerId = me
        let session = makeSession(for: me)

        let advertiser = MCNearbyServiceAdvertiser(
            peer: me,
            discoveryInfo: nil,
            serviceType: sanitize(serviceId)
        )
        advertiser.delegate = self
        advertiser.startAdvertisingPeer()
        self.advertiser = advertiser
        self.session = session
        call.resolve()
    }

    @objc func stopAdvertising(_ call: CAPPluginCall) {
        advertiser?.stopAdvertisingPeer()
        advertiser = nil
        call.resolve()
    }

    // MARK: - Découverte (invité)

    @objc func startDiscovery(_ call: CAPPluginCall) {
        guard let serviceId = call.getString("serviceId") else {
            call.reject("serviceId manquant")
            return
        }

        // Un pair qui ne fait que chercher a quand même besoin d'une identité
        // et d'une session : Multipeer n'a pas de mode « écoute seule ».
        let me = peerId ?? MCPeerID(displayName: UIDevice.current.name)
        peerId = me
        if session == nil { session = makeSession(for: me) }

        let browser = MCNearbyServiceBrowser(peer: me, serviceType: sanitize(serviceId))
        browser.delegate = self
        browser.startBrowsingForPeers()
        self.browser = browser
        call.resolve()
    }

    @objc func stopDiscovery(_ call: CAPPluginCall) {
        browser?.stopBrowsingForPeers()
        browser = nil
        call.resolve()
    }

    // MARK: - Connexion

    @objc func requestConnection(_ call: CAPPluginCall) {
        guard let endpointId = call.getString("endpointId"),
            let peer = found[endpointId],
            let session = session,
            let browser = browser
        else {
            call.reject("pair inconnu ou découverte non démarrée")
            return
        }
        browser.invitePeer(peer, to: session, withContext: nil, timeout: 20)
        call.resolve()
    }

    @objc func acceptConnection(_ call: CAPPluginCall) {
        guard let endpointId = call.getString("endpointId") else {
            call.reject("endpointId manquant")
            return
        }
        // L'acceptation est automatique côté socle : le code court a déjà servi
        // de filtre avant qu'on en arrive là.
        if let handler = pendingInvitations.removeValue(forKey: endpointId) {
            handler(true, session)
        }
        call.resolve()
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        // Multipeer ne sait pas couper un seul pair : `disconnect()` quitte la
        // session entière. À quatre joueurs maximum et une session par partie,
        // c'est sans conséquence — mais il faut le savoir.
        session?.disconnect()
        call.resolve()
    }

    @objc func send(_ call: CAPPluginCall) {
        guard let endpointId = call.getString("endpointId"),
            let base64 = call.getString("data"),
            let payload = Data(base64Encoded: base64),
            let session = session
        else {
            call.reject("endpointId ou data manquant")
            return
        }
        guard let peer = session.connectedPeers.first(where: { $0.displayName == endpointId })
        else {
            call.reject("pair non connecté : \(endpointId)")
            return
        }
        do {
            // `.reliable` : le socle suppose la remise et l'ordre, comme sur les
            // autres transports. Le netcode gère la perte à son niveau, pas ici.
            try session.send(payload, toPeers: [peer], with: .reliable)
            call.resolve()
        } catch {
            call.reject("envoi impossible : \(error.localizedDescription)")
        }
    }

    // MARK: - Outils

    private func makeSession(for peer: MCPeerID) -> MCSession {
        let session = MCSession(peer: peer, securityIdentity: nil, encryptionPreference: .required)
        session.delegate = self
        return session
    }

    /**
     Normalise un identifiant de service.

     Multipeer impose un format strict — 15 caractères au plus, minuscules,
     chiffres et tirets seulement — et **plante** si on lui donne autre chose.
     Le contrat, lui, laisse l'appelant libre : on filtre donc ici plutôt que
     de faire confiance.
     */
    private func sanitize(_ serviceId: String) -> String {
        let allowed = serviceId.lowercased().map { c -> Character in
            c.isLetter || c.isNumber || c == "-" ? c : "-"
        }
        return String(allowed.prefix(15))
    }
}

// MARK: - Annonce

extension NearbyPlugin: MCNearbyServiceAdvertiserDelegate {
    public func advertiser(
        _ advertiser: MCNearbyServiceAdvertiser,
        didReceiveInvitationFromPeer peerID: MCPeerID,
        withContext context: Data?,
        invitationHandler: @escaping (Bool, MCSession?) -> Void
    ) {
        let id = peerID.displayName
        found[id] = peerID
        pendingInvitations[id] = invitationHandler
        notifyListeners(
            "connectionRequested",
            data: ["endpointId": id, "endpointName": id])
    }

    public func advertiser(
        _ advertiser: MCNearbyServiceAdvertiser,
        didNotStartAdvertisingPeer error: Error
    ) {
        notifyListeners(
            "disconnected",
            data: ["endpointId": "", "reason": error.localizedDescription])
    }
}

// MARK: - Découverte

extension NearbyPlugin: MCNearbyServiceBrowserDelegate {
    public func browser(
        _ browser: MCNearbyServiceBrowser,
        foundPeer peerID: MCPeerID,
        withDiscoveryInfo info: [String: String]?
    ) {
        let id = peerID.displayName
        found[id] = peerID
        notifyListeners("endpointFound", data: ["endpointId": id, "endpointName": id])
    }

    public func browser(_ browser: MCNearbyServiceBrowser, lostPeer peerID: MCPeerID) {
        let id = peerID.displayName
        found.removeValue(forKey: id)
        notifyListeners("endpointLost", data: ["endpointId": id])
    }
}

// MARK: - Session

extension NearbyPlugin: MCSessionDelegate {
    public func session(
        _ session: MCSession,
        peer peerID: MCPeerID,
        didChange state: MCSessionState
    ) {
        let id = peerID.displayName
        switch state {
        case .connected:
            notifyListeners("connected", data: ["endpointId": id, "endpointName": id])
        case .notConnected:
            notifyListeners("disconnected", data: ["endpointId": id, "reason": "session fermée"])
        case .connecting:
            break
        @unknown default:
            break
        }
    }

    public func session(_ session: MCSession, didReceive data: Data, fromPeer peerID: MCPeerID) {
        notifyListeners(
            "received",
            data: [
                "endpointId": peerID.displayName,
                // Base64 : le pont Capacitor ne transporte que du JSON.
                "data": data.base64EncodedString(),
            ])
    }

    // Flux et ressources ne servent pas : le socle n'échange que de petits
    // messages, et un flux imposerait une gestion de découpage que le
    // `Channel` fait déjà, mieux et de façon commune à tous les transports.
    public func session(
        _ session: MCSession,
        didReceive stream: InputStream,
        withName streamName: String,
        fromPeer peerID: MCPeerID
    ) {}

    public func session(
        _ session: MCSession,
        didStartReceivingResourceWithName resourceName: String,
        fromPeer peerID: MCPeerID,
        with progress: Progress
    ) {}

    public func session(
        _ session: MCSession,
        didFinishReceivingResourceWithName resourceName: String,
        fromPeer peerID: MCPeerID,
        at localURL: URL?,
        withError error: Error?
    ) {}
}

File: native/capacitor-nearby/android/src/main/java/app/tictacdoh/nearby/NearbyPlugin.kt
package app.tictacdoh.nearby

import android.util.Base64
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.android.gms.nearby.Nearby
import com.google.android.gms.nearby.connection.AdvertisingOptions
import com.google.android.gms.nearby.connection.ConnectionInfo
import com.google.android.gms.nearby.connection.ConnectionLifecycleCallback
import com.google.android.gms.nearby.connection.ConnectionResolution
import com.google.android.gms.nearby.connection.ConnectionsStatusCodes
import com.google.android.gms.nearby.connection.DiscoveredEndpointInfo
import com.google.android.gms.nearby.connection.DiscoveryOptions
import com.google.android.gms.nearby.connection.EndpointDiscoveryCallback
import com.google.android.gms.nearby.connection.Payload
import com.google.android.gms.nearby.connection.PayloadCallback
import com.google.android.gms.nearby.connection.PayloadTransferUpdate
import com.google.android.gms.nearby.connection.Strategy

/**
 * Plugin « à proximité » pour Android, au-dessus de Nearby Connections.
 *
 * La stratégie `P2P_STAR` correspond exactement à la topologie du socle : un
 * hôte, jusqu'à trois invités, aucune liaison entre invités. Nearby monte du
 * Wi-Fi Direct quand il peut et retombe sur le Bluetooth sinon, sans que
 * l'application ait à le savoir.
 *
 * ⚠️ NON COMPILÉ NI TESTÉ. Aucun SDK Android ni appareil n'était disponible.
 * Le contrat honoré ici est figé par les tests de `packages/transport-nearby`.
 */
@CapacitorPlugin(name = "Nearby")
class NearbyPlugin : Plugin() {

    private val client by lazy { Nearby.getConnectionsClient(context) }

    /** Nom d'endpoint local, réutilisé à chaque demande de connexion. */
    private var localEndpointName: String = "ttd"

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        // Nearby Connections repose sur les services Google Play : absents des
        // appareils sans GMS, où ce transport ne fonctionnera jamais. Mieux
        // vaut le dire que d'échouer à la première connexion.
        val available = com.google.android.gms.common.GoogleApiAvailability.getInstance()
            .isGooglePlayServicesAvailable(context) == com.google.android.gms.common.ConnectionResult.SUCCESS
        call.resolve(
            JSObject()
                .put("available", available)
                .put("reason", if (available) "" else "services Google Play indisponibles"),
        )
    }

    @PluginMethod
    fun startAdvertising(call: PluginCall) {
        val serviceId = call.getString("serviceId") ?: return call.reject("serviceId manquant")
        val endpointName = call.getString("endpointName") ?: return call.reject("endpointName manquant")
        localEndpointName = endpointName

        client
            .startAdvertising(
                endpointName,
                serviceId,
                connectionLifecycle,
                AdvertisingOptions.Builder().setStrategy(Strategy.P2P_STAR).build(),
            )
            .addOnSuccessListener { call.resolve() }
            .addOnFailureListener { call.reject("advertising refusé: ${it.message}") }
    }

    @PluginMethod
    fun stopAdvertising(call: PluginCall) {
        client.stopAdvertising()
        call.resolve()
    }

    @PluginMethod
    fun startDiscovery(call: PluginCall) {
        val serviceId = call.getString("serviceId") ?: return call.reject("serviceId manquant")
        client
            .startDiscovery(
                serviceId,
                endpointDiscovery,
                DiscoveryOptions.Builder().setStrategy(Strategy.P2P_STAR).build(),
            )
            .addOnSuccessListener { call.resolve() }
            .addOnFailureListener { call.reject("découverte refusée: ${it.message}") }
    }

    @PluginMethod
    fun stopDiscovery(call: PluginCall) {
        client.stopDiscovery()
        call.resolve()
    }

    @PluginMethod
    fun requestConnection(call: PluginCall) {
        val endpointId = call.getString("endpointId") ?: return call.reject("endpointId manquant")
        val name = call.getString("endpointName") ?: localEndpointName
        client
            .requestConnection(name, endpointId, connectionLifecycle)
            .addOnSuccessListener { call.resolve() }
            .addOnFailureListener { call.reject("connexion refusée: ${it.message}") }
    }

    @PluginMethod
    fun acceptConnection(call: PluginCall) {
        val endpointId = call.getString("endpointId") ?: return call.reject("endpointId manquant")
        client
            .acceptConnection(endpointId, payloadCallback)
            .addOnSuccessListener { call.resolve() }
            .addOnFailureListener { call.reject("acceptation refusée: ${it.message}") }
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        val endpointId = call.getString("endpointId") ?: return call.reject("endpointId manquant")
        client.disconnectFromEndpoint(endpointId)
        call.resolve()
    }

    @PluginMethod
    fun send(call: PluginCall) {
        val endpointId = call.getString("endpointId") ?: return call.reject("endpointId manquant")
        val bytes = Base64.decode(call.getString("data") ?: "", Base64.NO_WRAP)
        // `Payload.fromBytes` est plafonné à 32 ko, ce qui correspond à la MTU
        // annoncée côté TypeScript : au-delà il faudrait un flux, dont le socle
        // n'a pas besoin puisque `Channel` fragmente déjà.
        client
            .sendPayload(endpointId, Payload.fromBytes(bytes))
            .addOnSuccessListener { call.resolve() }
            .addOnFailureListener { call.reject("envoi refusé: ${it.message}") }
    }

    private val endpointDiscovery = object : EndpointDiscoveryCallback() {
        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
            notifyListeners(
                "endpointFound",
                JSObject()
                    .put("endpointId", endpointId)
                    .put("endpointName", info.endpointName),
            )
        }

        override fun onEndpointLost(endpointId: String) {
            notifyListeners("endpointLost", JSObject().put("endpointId", endpointId))
        }
    }

    private val connectionLifecycle = object : ConnectionLifecycleCallback() {
        override fun onConnectionInitiated(endpointId: String, info: ConnectionInfo) {
            // Les deux bouts doivent accepter. Le côté TypeScript le fait
            // automatiquement : le code court a déjà servi de filtre, redemander
            // une confirmation à l'utilisateur n'apporterait rien.
            notifyListeners(
                "connectionRequested",
                JSObject()
                    .put("endpointId", endpointId)
                    .put("endpointName", info.endpointName),
            )
        }

        override fun onConnectionResult(endpointId: String, resolution: ConnectionResolution) {
            if (resolution.status.statusCode == ConnectionsStatusCodes.STATUS_OK) {
                notifyListeners(
                    "connected",
                    JSObject().put("endpointId", endpointId).put("endpointName", endpointId),
                )
            } else {
                notifyListeners(
                    "disconnected",
                    JSObject()
                        .put("endpointId", endpointId)
                        .put("reason", "connexion refusée (${resolution.status.statusCode})"),
                )
            }
        }

        override fun onDisconnected(endpointId: String) {
            notifyListeners("disconnected", JSObject().put("endpointId", endpointId))
        }
    }

    private val payloadCallback = object : PayloadCallback() {
        override fun onPayloadReceived(endpointId: String, payload: Payload) {
            val bytes = payload.asBytes() ?: return
            notifyListeners(
                "received",
                JSObject()
                    .put("endpointId", endpointId)
                    .put("data", Base64.encodeToString(bytes, Base64.NO_WRAP)),
            )
        }

        override fun onPayloadTransferUpdate(endpointId: String, update: PayloadTransferUpdate) {
            // Rien à faire : les charges utiles du socle tiennent toutes en un
            // seul morceau, la couche Channel ayant déjà fragmenté.
        }
    }
}
